[Publisher][Admin Panel] Add Vendor Management Modal

nasaownsky opened the pull request:
## Description of the change

Adding Vendor Management Modal UI and logic

## Type of change

> All pull requests must have at least one of the following labels applied (otherwise the PR will fail):

| Label                       	| Description                                                                                               	|
|-----------------------------	|-----------------------------------------------------------------------------------------------------------	|
| Type: Bug                   	| non-breaking change that fixes an issue                                                                   	|
| Type: Feature               	| non-breaking change that adds functionality                                                               	|
| Type: Breaking Change       	| fix or feature that would cause existing functionality to not work as expected                            	|
| Type: Non-breaking refactor 	| change addresses some tech debt item or prepares for a later change, but does not change functionality    	|
| Type: Configuration Change  	| adjusts configuration to achieve some end related to functionality, development, performance, or security 	|
| Type: Dependency Upgrade      | upgrades a project dependency - these changes are not included in release notes                             |

## Related issues

closes #1646 

## Checklists

### Development

**This box MUST be checked by the submitter prior to merging**:
- [ ] **Double- and triple-checked that there is no Personally Identifiable Information (PII) being mistakenly added in this pull request**

These boxes should be checked by the submitter prior to merging:
- [ ] Tests have been written to cover the code changed/added as part of this pull request

### Code review

These boxes should be checked by reviewers prior to merging:

- [ ] This pull request has a descriptive title and information useful to a reviewer
- [ ] This pull request has been moved out of a Draft state, has no "Work In Progress" label, and has assigned reviewers
- [ ] Potential security implications or infrastructural changes have been considered, if relevant

nasaownsky commented:
Hi @nichelle-hall @mxosman! A few question from the development:

1. The `admin/vendors` endpoint returns me HTML page with "You need to enable JavaScript to run this app." text in response. Is this expected because BE just isn't ready or something is wrong? Other endpoints (PUT and DELETE) also doesn't work with HTML "The method is not allowed for the requested URL" response.

2. In the Issue description there are requirements about Save button, and at first I was thinking about something like adding new or edited vendors to a new Set and the put in to BE with Save button, but as I can see from BE expected  [behavior](https://github.com/Recidiviz/justice-counts/issues/1646#issuecomment-2619773682) we can only add/edit one vendor at a time. Am I understanding correctly? In this case Add/Update button will be sending vendors straight off to BE, and I'm not exactly understand the purpose of the Save button.

3. Or we can just send something like this to PUT endpoint?

```
[
      { id: null, name: "Vendor A", url: "https://vendor-a.com" },
      { id: null, name: "Vendor B", url: "https://vendor-b.com" },
      { id: 1, name: "Vendor C", url: "https://vendor-c.com" },
      { id: 4, name: "Vendor D", url: "https://vendor-d.com" },
      { id: null, name: "Vendor E", url: "https://vendor-e.com" },
]
```
In this case Add/Update button will be adding/updating vendors in local Set and then sending it to BE via Save button.

nichelle-hall commented:
Hi @nasaownsky

1. The Endpoints:
I just merged all the backend endpoints as described in your Github Issue description. They will be available on staging tomorrow morning, around midday your time. In the meantime, you can point the FE to:
https://nichelletest---publisher-web-b47yvyxs3q-uc.a.run.app/

2. About the Save Button:
I may have interpreted the designs incorrectly. I initially understood that the "Add" button would immediately create the vendor through the backend, but I now see how this makes the "Save" button redundant. I’ll let @mxosman confirm the intended functionality and design. If the plan is to batch the additions/updates and only send them on "Save," I can update the endpoint to accept a list of vendors instead.

Let me know once you hear back!

mxosman commented:
Hi @nichelle-hall & @nasaownsky!

For question 2 - I was originally thinking the "Save" would send a batch update just for the sake of efficiency incase they need to add/edit/remove multiple vendors in one session. How cumbersome would it be to update the API to handle that @nichelle-hall? Sorry for not making this clear earlier. I also don't feel super strongly one-way or another, but figured it might be preferable to making a bunch of calls if users need to do a lot of vendor management.

nasaownsky commented:
The BE works as expected, thank you @nichelle-hall! 

@mxosman Hmm, if a user would create/add a bunch of vendors and then something goes wrong the progress can be lost, maybe it will be best to use current approach after all?  Anyway, please check my current implementation! 

1. Changed "Save" button to "Close"
2. Added "Clear" button 
3. Added borders for vendors to be better visually distinct

mxosman commented:
Hi @nasaownsky - I'm good with the current implementation! And, great call with the borders and buttons - it looks and feels great.

One small nit - if a user clicks Edit and then removes the vendor, can we clear the input? Just for the scenario where they edit an input they removed. If they're adding and click remove, we should keep their input. LMK if that's a huge pain.

https://github.com/user-attachments/assets/4f17ba3f-a273-496b-b548-08350960b102

Will do a code review later on this afternoon.

mxosman commented on publisher/src/components/AdminPanel/AgencyProvisioningOverview.tsx:
```diff
@@ -285,8 +294,14 @@ export const AgencyProvisioningOverview = observer(() => {
-        {/* Create Agency Button */}
         <Styled.ButtonWrapper>
+          {/* Manage Vendors Button */}
+          <Button
```
Can we gate this button to staging and dev environments for now?

mxosman commented on publisher/src/components/AdminPanel/VendorManagementModal.tsx:
```diff
@@ -0,0 +1,195 @@
+      !urlValidationError &&
+      (nameValue !== editNameValue || urlValue !== editlUrlValue); // Changes detected in "Update" mode
+
+    useEffect(() => {
+      if (!nameValue && !urlValue) {
+        handleClearVendorInfo();
+      }
+    }, [nameValue, urlValue]);
```
Can you help me understand what's going on with this effect? I'm not 100% following. Why are we clearing vendor info if it's already pretty much cleared? Or is it to clear the other states (editNameValue, and editURLValue, and UrlValidationError)?

nasaownsky replied:
If you comment out this effect, then go to Manage Vendors, hit edit and then manually delete Name and URL the modal will remain in "edit" state with the button showing "Update" text. I feel like this is kinda not right, and when you deleting these values you should go back to "add" mode, so this effect handles that.

mxosman replied:
OH - I see. That makes a lot of sense. Thank you for helping me understand! Great call on that detail!

mxosman commented on publisher/src/stores/AdminPanelStore.ts:
```diff
@@ -631,7 +695,8 @@ class AdminPanelStore {
       | UserWithAgenciesByID
       | AgencyWithTeamByID
-      | AgencyTeamMember
+      // eslint-disable-next-line prettier/prettier
```
Oi - any idea what's been going on with our `prettier` lately?

nasaownsky replied:
I believe there are conflicts between our `@recidiviz/eslint-config` and default prettier config files. I see that `@recidiviz/eslint-config` is not presented in `recidiviz-dashboards` repo and I was not observing prettier problems there in my time.

mxosman replied:
I'm totally good removing it from our repo - esp. if `recidiviz-dashbaords` no longer uses it. Sounds like it needs some updating and folks felt it wasn't worth it. Feel free to do it in this PR or a follow up one!

nasaownsky replied:
@mxosman Hmm, I'm played with it a bit and it seems like if we remove this from our repo -- it will not have `eslint` configured at all. What do you think?

mxosman replied:
Oh strange - I would assume it would fallback to the default eslint package, no? We probably have to update a few more configurations to delink from the @recidiviz/eslint-config - we can worry about it in a separate PR. I'll look into it or write up a ticket. Thanks for entertaining it!

nasaownsky replied:
Got it, thank you!

mxosman commented on publisher/src/components/AdminPanel/VendorManagementModal.tsx:
```diff
@@ -0,0 +1,195 @@
+        </Styled.VendorsWrapper>
+      </Modal>
+    );
+  });
```
So clean!

nasaownsky replied:
Thank you!

mxosman reviewed (commented):
Code looks great to me! Made one suggestion to gate it and another in my previous comment - otherwise, this is practically good to go!

mxosman reviewed (approved):
Amazing, as always! Thank you so much, @nasaownsky!
